src/RamTransforms.cpp

```diff
@@ -0,0 +1,432 @@
+                if (const RamScan* scan = dynamic_cast<const RamScan*>(&nested->getOperation())) {
+                    if (std::unique_ptr<RamOperation> op = rewriteScan(scan)) {
+                        modified = true;
+                        nested->setOperation(std::move(op));
+                    }
+                }
+            }
+
+            node->apply(*this);
+            return node;
+        }
+    };
+
+    // Node-mapper that searches for and updates RAM inserts
+    struct RamInsertCapturer : public RamNodeMapper {
+        mutable bool modified = false;
+
+        bool getModified() const {
+            return modified;
+        }
+
+        std::unique_ptr<RamNode> operator()(std::unique_ptr<RamNode> node) const override {
+            // get all RAM inserts
+            if (RamInsert* insert = dynamic_cast<RamInsert*>(node.get())) {
+                // TODO: better way to modify the child of a RAM insert
+                if (const RamScan* scan = dynamic_cast<const RamScan*>(&insert->getOperation())) {
+                    if (std::unique_ptr<RamOperation> op = rewriteScan(scan)) {
+                        modified = true;
+                        insert->setOperation(std::move(op));
+                    }
+                }
+
+                RamScanCapturer scanUpdate;
+                insert->apply(scanUpdate);
+
+                if (!modified && scanUpdate.getModified()) {
+                    modified = true;
+                }
+            } else {
+                // no need to search for nested RAM inserts
+                node->apply(*this);
+            }
+
+            return node;
+        }
+    };
+
+    // level all RAM inserts
+    RamInsertCapturer insertUpdate;
+    program.getMain()->apply(insertUpdate);
+
+    return insertUpdate.getModified();
+}
+
+bool ConvertExistenceChecksTransformer::convertExistenceChecks(RamProgram& program) {
+    // TODO: Change these to LambdaRamNodeMapper lambdas
+    // Node-mapper that searches for and updates RAM scans nested in RAM inserts
+    struct RamScanCapturer : public RamNodeMapper {
+        mutable bool modified = false;
+
+        bool getModified() const {
+            return modified;
+        }
+
+        bool dependsOn(const RamCondition* condition, const size_t identifier) const {
+            if (const RamBinaryRelation* binRel = dynamic_cast<const RamBinaryRelation*>(condition)) {
+                std::vector<const RamValue*> queue = {binRel->getLHS(), binRel->getRHS()};
+                while (!queue.empty()) {
+                    const RamValue* val = queue.back();
+                    queue.pop_back();
+                    if (const RamValue* elemAccess = dynamic_cast<const RamElementAccess*>(val)) {
+                        if (elemAccess->getLevel() == identifier) {
+                            return true;
+                        }
+                    } else if (const RamBinaryOperator* binOp = dynamic_cast<const RamBinaryOperator*>(val)) {
```

Comment:
Do other cases need to be covered (e.g. ternary operators, user-defined operators, etc.)?

Reply:
Thanks this is a good point. This function is also quite important because if it misses a tuple reference then it may convert a scan to an existence check where a scan was required. Ideally with refactored functors this will become more unified than having to deal with each case separately.